event_sale_seat_reserve/models/sale_order_line.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl-3.0).
+import logging
+
+from odoo import models
+from odoo.tools import config
+
+_logger = logging.getLogger(__name__)
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    def _update_registrations(
+        self,
+        confirm=True,
+        cancel_to_draft=False,
+        registration_data=None,
+        mark_as_paid=False,
+    ):
+        res = super(SaleOrderLine, self)._update_registrations(
+            confirm=confirm,
+            cancel_to_draft=cancel_to_draft,
+            registration_data=registration_data,
+            mark_as_paid=mark_as_paid,
+        )
+        RegistrationSudo = self.env["event.registration"].sudo()
+        registrations = RegistrationSudo.search(
+            [("sale_order_line_id", "in", self.ids), ("state", "=", "draft")]
+        )
+        if config["test_enable"] and not self.env.context.get(
+            "test_event_seat_reserve"
+        ):
+            _logger.info("Test mode is enabled, skipping the reservation of the seats")
+            return res
+        registrations.action_set_reserved()
+        return res
+
+    def _set_draft_associated_registrations(self):
+        RegistrationSudo = self.env["event.registration"].sudo()
+        registrations = RegistrationSudo.search(
+            [("sale_order_line_id", "in", self.ids), ("state", "!=", "done")]
+        )
+        registrations.action_set_draft()
+        return True
```

Comment:
nitpicking: I'd narrow the scope to the reserved registrations only
and perhaps rename the method to reflect that

```suggestion
    def _unreserve_registrations(self):
        RegistrationSudo = self.env["event.registration"].sudo()
        registrations = RegistrationSudo.search(
            [("sale_order_line_id", "in", self.ids), ("state", "=", "reserved")]
        )
        registrations.action_set_draft()
        return True
```

Reply:
This method is called from "Set to quotation" action.
Normally the button "Set to quotation" is displayed only after cancel the sale.
When cancelling a sale, all registrations linked to the sale lines will be cancelled too.

So here I'm trying to "re-activate" the linked registration : 

Sale order cancelled > sale order lines cancelled > registrations cancelled
Sale order set to draft > sale order lines set to draft > registrations set to reserved

I'm using `!= done` because it's possible that the user was changed manually the registration line to draft.